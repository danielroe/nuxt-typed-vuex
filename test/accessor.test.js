const { useAccessor } = require('../')
const vuex = require('vuex')
const Vue = require('vue')
const {
  getters,
  state,
  actions,
  mutations,
  modules,
} = require('./fixture/store')

const submodule = require('./fixture/store/submodule')

const pattern = {
  getters,
  state,
  actions,
  mutations,
  modules: {
    submodule: {
      ...submodule,
      namespaced: true,
      modules: {
        nestedSubmodule: {
          ...submodule,
          namespaced: true,
        }
      }
    },
  },
}

describe.only('accessor', () => {
  let store
  let accessor
  beforeEach(() => {
    Vue.use(vuex)
    store = new vuex.Store(pattern)
    accessor = useAccessor(store, pattern)
  })
  test('store exists', async () => {
    expect(pattern).toMatchSnapshot()
  })
  test('accessor object exists', async () => {
    expect(accessor).toMatchSnapshot()
  })
  test('accessor state, getter, mutation and actions work', async () => {
    expect(accessor.fullEmail).toEqual('')
    accessor.setEmail('john@j.com')
    expect(accessor.email).toEqual('john@j.com')
    expect(accessor.fullEmail).toEqual('john@j.com')
    accessor.resetEmail()
    expect(accessor.fullEmail).toEqual('a@a.com')
  })
  test('accessor submodule state, getter, mutation and actions work', async () => {
    expect(accessor.submodule.firstName).toEqual('')
    accessor.submodule.setFirstName('Nina')
    expect(accessor.submodule.firstName).toEqual('Nina')
    accessor.submodule.setLastName('Willis')
    expect(accessor.submodule.fullName).toEqual('Nina Willis')
    accessor.submodule.initialise()
    expect(accessor.submodule.fullName).toEqual('John Baker')
    accessor.submodule.setName('Jordan Lawrence')
    expect(accessor.submodule.firstName).toEqual('Jordan')
  })
  test('nested modules work', async () => {
    expect(accessor.submodule.nestedSubmodule.firstName).toEqual('')
    accessor.submodule.nestedSubmodule.setFirstName('Nina')
    expect(accessor.submodule.nestedSubmodule.firstName).toEqual('Nina')
    accessor.submodule.nestedSubmodule.setLastName('Willis')
    expect(accessor.submodule.nestedSubmodule.fullName).toEqual('Nina Willis')
    accessor.submodule.nestedSubmodule.initialise()
    expect(accessor.submodule.nestedSubmodule.fullName).toEqual('John Baker')
    accessor.submodule.nestedSubmodule.setName('Jordan Lawrence')
    expect(accessor.submodule.nestedSubmodule.firstName).toEqual('Jordan')
  })
  test('dynamic modules work', async () => {
    store.registerModule('submodule', submodule)
    const dynamicAccessor = useAccessor(store, { modules: { submodule } })

    expect(dynamicAccessor.submodule.firstName).toEqual('')
    dynamicAccessor.submodule.setFirstName('Nina')
    expect(dynamicAccessor.submodule.firstName).toEqual('Nina')
    dynamicAccessor.submodule.setLastName('Willis')
    expect(dynamicAccessor.submodule.fullName).toEqual('Nina Willis')
    dynamicAccessor.submodule.initialise()
    expect(dynamicAccessor.submodule.fullName).toEqual('John Baker')
    dynamicAccessor.submodule.setName('Jordan Lawrence')
    expect(dynamicAccessor.submodule.firstName).toEqual('Jordan')
  })
  test('dynamic module helper function works', async () => {
    store.registerModule('submodule', submodule)
    const dynamicAccessor = useAccessor(store, submodule, 'submodule')

    expect(dynamicAccessor.firstName).toEqual('')
    dynamicAccessor.setFirstName('Nina')
    expect(dynamicAccessor.firstName).toEqual('Nina')
    dynamicAccessor.setLastName('Willis')
    expect(dynamicAccessor.fullName).toEqual('Nina Willis')
    dynamicAccessor.initialise()
    expect(dynamicAccessor.fullName).toEqual('John Baker')
    dynamicAccessor.setName('Jordan Lawrence')
    expect(dynamicAccessor.firstName).toEqual('Jordan')
  })
})
