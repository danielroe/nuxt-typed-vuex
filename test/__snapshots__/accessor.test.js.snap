// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accessor accessor object exists 1`] = `
Object {
  "initialiseStore": [Function],
  "resetEmail": [Function],
  "setEmail": [Function],
  "submodule": Object {
    "initialise": [Function],
    "nestedSubmodule": Object {
      "initialise": [Function],
      "setFirstName": [Function],
      "setLastName": [Function],
      "setName": [Function],
    },
    "setFirstName": [Function],
    "setLastName": [Function],
    "setName": [Function],
  },
}
`;

exports[`accessor store exists 1`] = `
Object {
  "actions": Object {
    "resetEmail": [Function],
  },
  "getters": Object {
    "email": [Function],
    "fullEmail": [Function],
  },
  "modules": Object {
    "submodule": Object {
      "actions": Object {
        "initialise": [Function],
        "setName": [Function],
      },
      "getters": Object {
        "fullName": [Function],
      },
      "modules": Object {
        "nestedSubmodule": Object {
          "actions": Object {
            "initialise": [Function],
            "setName": [Function],
          },
          "getters": Object {
            "fullName": [Function],
          },
          "mutations": Object {
            "setFirstName": [Function],
            "setLastName": [Function],
          },
          "namespaced": true,
          "state": [Function],
        },
      },
      "mutations": Object {
        "setFirstName": [Function],
        "setLastName": [Function],
      },
      "namespaced": true,
      "state": [Function],
    },
  },
  "mutations": Object {
    "initialiseStore": [Function],
    "setEmail": [Function],
  },
  "state": [Function],
}
`;
